prow/plank/controller.go

```diff
@@ -0,0 +1,275 @@
+		crierURL: crierURL,
+		totURL:   totURL,
+	}
+}
+
+func (c *Controller) Sync() error {
+	pjs, err := c.kc.ListProwJobs(nil)
+	if err != nil {
+		return fmt.Errorf("error listing prow jobs: %v", err)
+	}
+	pods, err := c.kc.ListPods(nil)
+	if err != nil {
+		return fmt.Errorf("error listing pods: %v", err)
+	}
+	pm := map[string]kube.Pod{}
+	for _, pod := range pods {
+		pm[pod.Metadata.Name] = pod
+	}
+	var errs []error
+	for _, pj := range pjs {
+		if err := c.syncJob(pj, pm); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) == 0 {
+		return nil
+	} else {
+		return fmt.Errorf("errors syncing: %v", errs)
+	}
+}
+
+func (c *Controller) syncJob(pj kube.ProwJob, pm map[string]kube.Pod) error {
+	var updated bool
+	if pj.Complete() {
+		// ProwJob is complete. Do nothing.
+	} else if pj.Status.PodName == "" {
+		// We haven't started the pod yet. Do so.
+		updated = true
+		pj.Status.State = kube.PendingState
+		if pn, err := c.startPod(pj); err == nil {
+			pj.Status.PodName = pn
+		} else {
+			return fmt.Errorf("error starting pod: %v", err)
+		}
+		if err := c.report(pj); err != nil {
+			return fmt.Errorf("error reporting to crier: %v", err)
+		}
+	} else if pod, ok := pm[pj.Status.PodName]; !ok {
+		// Pod is missing. This shouldn't happen normally, but if someone goes
+		// in and manually deletes the pod then we'll hit it. Start a new pod.
+		updated = true
+		pj.Status.PodName = ""
+		pj.Status.State = kube.PendingState
+	} else if pod.Status.Phase == kube.PodUnknown {
+		// Pod is in Unknown state. This can happen if there is a problem with
+		// the node. Delete the old pod, we'll start a new one next loop.
+		updated = true
+		if err := c.kc.DeletePod(pj.Status.PodName); err != nil {
+			return fmt.Errorf("error deleting pod %s: %v", pj.Status.PodName, err)
+		}
+		pj.Status.PodName = ""
+		pj.Status.State = kube.PendingState
+	} else if pod.Status.Phase == kube.PodSucceeded {
+		// Pod succeeded. Update ProwJob, talk to crier, and start next jobs.
+		updated = true
+		pj.Status.CompletionTime = time.Now()
+		pj.Status.State = kube.SuccessState
+		if err := c.report(pj); err != nil {
+			return fmt.Errorf("error reporting to crier: %v", err)
+		}
+		for _, nj := range pj.Spec.RunAfterSuccess {
+			if _, err := c.kc.CreateProwJob(NewProwJob(nj)); err != nil {
+				return fmt.Errorf("error starting next prowjob: %v", err)
+			}
+		}
+	} else if pod.Status.Phase == kube.PodFailed {
+		// Pod failed. Update ProwJob, talk to crier.
+		updated = true
+		pj.Status.CompletionTime = time.Now()
+		pj.Status.State = kube.FailureState
+		if err := c.report(pj); err != nil {
+			return fmt.Errorf("error reporting to crier: %v", err)
+		}
+	} else {
+		// Pod is running. Do nothing.
+	}
+	if updated {
+		_, err := c.kc.ReplaceProwJob(pj.Metadata.Name, pj)
+		return err
+	} else {
+		return nil
+	}
+}
+
+func (c *Controller) report(pj kube.ProwJob) error {
+	if !pj.Spec.Report {
+		return nil
+	}
+	if len(pj.Spec.Refs.Pulls) != 1 {
```

Comment:
It only makes sense to report to GitHub for the case where we have one PR. Postsubmits and Periodics have <1, batches have >1.